

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>P1_utils &mdash; CANDLE  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> CANDLE
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../readme.html">What is CANDLE?</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/index.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../candle_lib.html">CANDLE Library API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contribute.html">How to Contribute</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">CANDLE</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="index.html">Module code</a> &raquo;</li>
        
      <li>P1_utils</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for P1_utils</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">patsy</span>
<span class="kn">import</span> <span class="nn">numpy.linalg</span> <span class="k">as</span> <span class="nn">la</span>
<span class="kn">from</span> <span class="nn">sklearn.feature_selection</span> <span class="k">import</span> <span class="n">mutual_info_regression</span>
<span class="kn">import</span> <span class="nn">statsmodels.api</span> <span class="k">as</span> <span class="nn">sm</span>


<span class="c1">################### Auxiliary functions of COXEN start here ####################</span>



<div class="viewcode-block" id="calculate_concordance_correlation_coefficient"><a class="viewcode-back" href="../candle_lib/P1_utils.html#P1_utils.calculate_concordance_correlation_coefficient">[docs]</a><span class="k">def</span> <span class="nf">calculate_concordance_correlation_coefficient</span><span class="p">(</span><span class="n">u</span><span class="p">,</span> <span class="n">v</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    This function calculates the concordance correlation coefficient between two input 1-D numpy arrays.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    u: 1-D numpy array of a variable</span>
<span class="sd">    v: 1-D numpy array of a variable</span>

<span class="sd">    Returns:</span>
<span class="sd">    --------</span>
<span class="sd">    ccc: a numeric value of concordance correlation coefficient between the two input variables.</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">a</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">((</span><span class="n">u</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">u</span><span class="p">))</span> <span class="o">*</span> <span class="p">(</span><span class="n">v</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">v</span><span class="p">)))</span>
    <span class="n">b</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">square</span><span class="p">(</span><span class="n">u</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">u</span><span class="p">)))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">square</span><span class="p">(</span><span class="n">v</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">v</span><span class="p">)))</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">square</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">u</span><span class="p">)</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">v</span><span class="p">))</span>
    <span class="n">ccc</span> <span class="o">=</span> <span class="n">a</span><span class="o">/</span><span class="n">b</span>
    <span class="k">return</span> <span class="n">ccc</span></div>



<div class="viewcode-block" id="generalization_feature_selection"><a class="viewcode-back" href="../candle_lib/P1_utils.html#P1_utils.generalization_feature_selection">[docs]</a><span class="k">def</span> <span class="nf">generalization_feature_selection</span><span class="p">(</span><span class="n">data1</span><span class="p">,</span> <span class="n">data2</span><span class="p">,</span> <span class="n">measure</span><span class="p">,</span> <span class="n">cutoff</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    This function uses the Pearson correlation coefficient to select the features that are generalizable</span>
<span class="sd">    between data1 and data2.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    data1: 2D numpy array of the first dataset with a size of (n_samples_1, n_features)</span>
<span class="sd">    data2: 2D numpy array of the second dataset with a size of (n_samples_2, n_features)</span>
<span class="sd">    measure: string. &#39;pearson&#39; indicates the Pearson correlation coefficient;</span>
<span class="sd">        &#39;ccc&#39; indicates the concordance correlation coefficient. Default is &#39;pearson&#39;.</span>
<span class="sd">    cutoff: a positive number for selecting generalizable features. If cutoff &lt; 1, this function selects</span>
<span class="sd">        the features with a correlation coefficient &gt;= cutoff. If cutoff &gt;= 1, it must be an</span>
<span class="sd">        integer indicating the number of features to be selected based on correlation coefficient.</span>

<span class="sd">    Returns:</span>
<span class="sd">    --------</span>
<span class="sd">    fid: 1-D numpy array containing the indices of selected features.</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">cor1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">corrcoef</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">transpose</span><span class="p">(</span><span class="n">data1</span><span class="p">))</span>
    <span class="n">cor2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">corrcoef</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">transpose</span><span class="p">(</span><span class="n">data2</span><span class="p">))</span>
    <span class="n">num</span> <span class="o">=</span> <span class="n">data1</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">cor</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">measure</span> <span class="o">==</span> <span class="s1">&#39;pearson&#39;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num</span><span class="p">):</span>
            <span class="n">cor</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">corrcoef</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">((</span><span class="nb">list</span><span class="p">(</span><span class="n">cor1</span><span class="p">[:</span><span class="n">i</span><span class="p">,</span> <span class="n">i</span><span class="p">])</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">cor1</span><span class="p">[(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):,</span> <span class="n">i</span><span class="p">]),</span>
                       <span class="nb">list</span><span class="p">(</span><span class="n">cor2</span><span class="p">[:</span><span class="n">i</span><span class="p">,</span> <span class="n">i</span><span class="p">])</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">cor2</span><span class="p">[(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):,</span> <span class="n">i</span><span class="p">]))))[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
    <span class="k">elif</span> <span class="n">measure</span> <span class="o">==</span> <span class="s1">&#39;ccc&#39;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num</span><span class="p">):</span>
            <span class="n">cor</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">calculate_concordance_correlation_coefficient</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">cor1</span><span class="p">[:</span><span class="n">i</span><span class="p">,</span> <span class="n">i</span><span class="p">])</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">cor1</span><span class="p">[(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):,</span> <span class="n">i</span><span class="p">])),</span>
                <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">cor2</span><span class="p">[:</span><span class="n">i</span><span class="p">,</span> <span class="n">i</span><span class="p">])</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">cor2</span><span class="p">[(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):,</span> <span class="n">i</span><span class="p">]))))</span>
    <span class="n">cor</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">cor</span><span class="p">)</span>
    <span class="n">fid</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="n">cor</span><span class="p">)[:</span><span class="nb">int</span><span class="p">(</span><span class="n">cutoff</span><span class="p">)]</span>
    <span class="k">return</span> <span class="n">fid</span></div>



<span class="c1">################### Auxiliary functions of COXEN end here ####################</span>

<div class="viewcode-block" id="coxen_single_drug_gene_selection"><a class="viewcode-back" href="../candle_lib/P1_utils.html#P1_utils.coxen_single_drug_gene_selection">[docs]</a><span class="k">def</span> <span class="nf">coxen_single_drug_gene_selection</span><span class="p">(</span><span class="n">source_data</span><span class="p">,</span> <span class="n">target_data</span><span class="p">,</span> <span class="n">drug_response_data</span><span class="p">,</span> <span class="n">drug_response_col</span><span class="p">,</span> <span class="n">tumor_col</span><span class="p">,</span>
    <span class="n">prediction_power_measure</span><span class="o">=</span><span class="s1">&#39;pearson&#39;</span><span class="p">,</span> <span class="n">num_predictive_gene</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">generalization_power_measure</span><span class="o">=</span><span class="s1">&#39;ccc&#39;</span><span class="p">,</span>
    <span class="n">num_generalizable_gene</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">multi_drug_mode</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    This function selects genes for drug response prediction using the COXEN approach. The COXEN approach is</span>
<span class="sd">    designed for selecting genes to predict the response of tumor cells to a specific drug. This function</span>
<span class="sd">    assumes no missing data exist.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    source_data: pandas data frame of gene expressions of tumors, for which drug response is known. Its size is</span>
<span class="sd">        [n_source_samples, n_features].</span>
<span class="sd">    target_data: pandas data frame of gene expressions of tumors, for which drug response needs to be predicted.</span>
<span class="sd">        Its size is [n_target_samples, n_features]. source_data and target_data have the same set</span>
<span class="sd">        of features and the orders of features must match.</span>
<span class="sd">    drug_response_data: pandas data frame of drug response values for a drug. It must include a column of drug</span>
<span class="sd">        response values and a column of tumor IDs.</span>
<span class="sd">    drug_response_col: non-negative integer or string. If integer, it is the column index of drug response in</span>
<span class="sd">        drug_response_data. If string, it is the column name of drug response.</span>
<span class="sd">    tumor_col: non-negative integer or string. If integer, it is the column index of tumor IDs in drug_response_data.</span>
<span class="sd">        If string, it is the column name of tumor IDs.</span>
<span class="sd">    prediction_power_measure: string. &#39;pearson&#39; uses the absolute value of Pearson correlation coefficient to</span>
<span class="sd">        measure prediction power of gene; &#39;mutual_info&#39; uses the mutual information to measure prediction power</span>
<span class="sd">        of gene. Default is &#39;pearson&#39;.</span>
<span class="sd">    num_predictive_gene: positive integer indicating the number of predictive genes to be selected.</span>
<span class="sd">    generalization_power_measure: string. &#39;pearson&#39; indicates the Pearson correlation coefficient;</span>
<span class="sd">        &#39;ccc&#39; indicates the concordance correlation coefficient. Default is &#39;ccc&#39;.</span>
<span class="sd">    num_generalizable_gene: positive integer indicating the number of generalizable genes to be selected.</span>
<span class="sd">    multi_drug_mode: boolean, indicating whether the function runs as an auxiliary function of COXEN</span>
<span class="sd">        gene selection for multiple drugs. Default is False.</span>

<span class="sd">    Returns:</span>
<span class="sd">    --------</span>
<span class="sd">    indices: 1-D numpy array containing the indices of selected genes, if multi_drug_mode is False;</span>
<span class="sd">    1-D numpy array of indices of sorting all genes according to their prediction power, if multi_drug_mode is True.</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">drug_response_col</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">drug_response_col</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">drug_response_data</span><span class="o">.</span><span class="n">columns</span> <span class="o">==</span> <span class="n">drug_response_col</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">tumor_col</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">tumor_col</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">drug_response_data</span><span class="o">.</span><span class="n">columns</span> <span class="o">==</span> <span class="n">tumor_col</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

    <span class="n">drug_response_data</span> <span class="o">=</span> <span class="n">drug_response_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">drug_response_data</span> <span class="o">=</span> <span class="n">drug_response_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">drug_response_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">tumor_col</span><span class="p">],</span>
                                                                  <span class="n">source_data</span><span class="o">.</span><span class="n">index</span><span class="p">))[</span><span class="mi">0</span><span class="p">],</span> <span class="p">:]</span>

    <span class="n">source_data</span> <span class="o">=</span> <span class="n">source_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">source_data</span> <span class="o">=</span> <span class="n">source_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">source_data</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">drug_response_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">tumor_col</span><span class="p">]))[</span><span class="mi">0</span><span class="p">],</span> <span class="p">:]</span>

    <span class="n">source_std_id</span> <span class="o">=</span> <span class="n">select_features_by_variation</span><span class="p">(</span><span class="n">source_data</span><span class="p">,</span> <span class="n">variation_measure</span><span class="o">=</span><span class="s1">&#39;std&#39;</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">0.00000001</span><span class="p">)</span>
    <span class="n">target_std_id</span> <span class="o">=</span> <span class="n">select_features_by_variation</span><span class="p">(</span><span class="n">target_data</span><span class="p">,</span> <span class="n">variation_measure</span><span class="o">=</span><span class="s1">&#39;std&#39;</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">0.00000001</span><span class="p">)</span>
    <span class="n">std_id</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="n">source_std_id</span><span class="p">,</span> <span class="n">target_std_id</span><span class="p">))</span>
    <span class="n">source_data</span> <span class="o">=</span> <span class="n">source_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">std_id</span><span class="p">]</span>
    <span class="n">target_data</span> <span class="o">=</span> <span class="n">target_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">target_data</span> <span class="o">=</span> <span class="n">target_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">std_id</span><span class="p">]</span>

    <span class="c1"># Perform the first step of COXEN approach to select predictive genes. To avoid exceeding the memory limit,</span>
    <span class="c1"># the prediction power of genes is calculated in batches.</span>
    <span class="n">batchSize</span> <span class="o">=</span> <span class="mi">1000</span>
    <span class="n">numBatch</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="n">source_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">/</span><span class="n">batchSize</span><span class="p">))</span>
    <span class="n">prediction_power</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">((</span><span class="n">source_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="mi">1</span><span class="p">))</span>
    <span class="n">prediction_power</span><span class="o">.</span><span class="n">fill</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">numBatch</span><span class="p">):</span>
        <span class="n">startIndex</span> <span class="o">=</span> <span class="n">i</span><span class="o">*</span><span class="n">batchSize</span>
        <span class="n">endIndex</span> <span class="o">=</span> <span class="nb">min</span><span class="p">((</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">*</span><span class="n">batchSize</span><span class="p">,</span> <span class="n">source_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

        <span class="k">if</span> <span class="n">prediction_power_measure</span> <span class="o">==</span> <span class="s1">&#39;pearson&#39;</span><span class="p">:</span>
            <span class="n">cor_i</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">corrcoef</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">transpose</span><span class="p">(</span><span class="n">source_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">startIndex</span><span class="p">:</span><span class="n">endIndex</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">drug_response_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">tumor_col</span><span class="p">],</span>
                <span class="p">:]</span><span class="o">.</span><span class="n">values</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">drug_response_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">drug_response_col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">drug_response_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])))))</span>
            <span class="n">prediction_power</span><span class="p">[</span><span class="n">startIndex</span><span class="p">:</span><span class="n">endIndex</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">cor_i</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">])</span>

        <span class="k">if</span> <span class="n">prediction_power_measure</span> <span class="o">==</span> <span class="s1">&#39;mutual_info&#39;</span><span class="p">:</span>
            <span class="n">mi</span> <span class="o">=</span> <span class="n">mutual_info_regression</span><span class="p">(</span><span class="n">X</span><span class="o">=</span><span class="n">source_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">startIndex</span><span class="p">:</span><span class="n">endIndex</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">drug_response_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">tumor_col</span><span class="p">],</span> <span class="p">:]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
                                <span class="n">y</span><span class="o">=</span><span class="n">drug_response_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">drug_response_col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
            <span class="n">prediction_power</span><span class="p">[</span><span class="n">startIndex</span><span class="p">:</span><span class="n">endIndex</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">mi</span>

    <span class="k">if</span> <span class="n">multi_drug_mode</span><span class="p">:</span>
        <span class="n">indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="n">prediction_power</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">std_id</span><span class="p">[</span><span class="n">indices</span><span class="p">]</span>

    <span class="n">num_predictive_gene</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">num_predictive_gene</span><span class="p">,</span> <span class="n">source_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
    <span class="n">gid1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="n">prediction_power</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">])[:</span><span class="n">num_predictive_gene</span><span class="p">]</span>

    <span class="c1"># keep only predictive genes for source and target data</span>
    <span class="n">source_data</span> <span class="o">=</span> <span class="n">source_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">gid1</span><span class="p">]</span>
    <span class="n">target_data</span> <span class="o">=</span> <span class="n">target_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">gid1</span><span class="p">]</span>
    <span class="n">num_generalizable_gene</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">num_generalizable_gene</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">gid1</span><span class="p">)))</span>
    <span class="c1"># perform the second step of COXEN approach to select generalizable genes among the predictive genes</span>
    <span class="n">gid2</span> <span class="o">=</span> <span class="n">generalization_feature_selection</span><span class="p">(</span><span class="n">source_data</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">target_data</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">generalization_power_measure</span><span class="p">,</span>
                                            <span class="n">num_generalizable_gene</span><span class="p">)</span>

    <span class="n">indices</span> <span class="o">=</span> <span class="n">std_id</span><span class="p">[</span><span class="n">gid1</span><span class="p">[</span><span class="n">gid2</span><span class="p">]]</span>

    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span></div>

<div class="viewcode-block" id="coxen_multi_drug_gene_selection"><a class="viewcode-back" href="../candle_lib/P1_utils.html#P1_utils.coxen_multi_drug_gene_selection">[docs]</a><span class="k">def</span> <span class="nf">coxen_multi_drug_gene_selection</span><span class="p">(</span><span class="n">source_data</span><span class="p">,</span> <span class="n">target_data</span><span class="p">,</span> <span class="n">drug_response_data</span><span class="p">,</span> <span class="n">drug_response_col</span><span class="p">,</span> <span class="n">tumor_col</span><span class="p">,</span> <span class="n">drug_col</span><span class="p">,</span>
                         <span class="n">prediction_power_measure</span><span class="o">=</span><span class="s1">&#39;lm&#39;</span><span class="p">,</span> <span class="n">num_predictive_gene</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">generalization_power_measure</span><span class="o">=</span><span class="s1">&#39;ccc&#39;</span><span class="p">,</span>
                         <span class="n">num_generalizable_gene</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">union_of_single_drug_selection</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    This function uses the COXEN approach to select genes for predicting the response of multiple drugs.</span>
<span class="sd">    It assumes no missing data exist. It works in three modes.</span>
<span class="sd">    (1) If union_of_single_drug_selection is True, prediction_power_measure must be either &#39;pearson&#39; or &#39;mutual_info&#39;.</span>
<span class="sd">    This functions runs coxen_single_drug_gene_selection for every drug with the parameter setting and takes the</span>
<span class="sd">    union of the selected genes of every drug as the output. The size of the selected gene set may be larger than</span>
<span class="sd">    num_generalizable_gene.</span>
<span class="sd">    (2) If union_of_single_drug_selection is False and prediction_power_measure is &#39;lm&#39;, this function uses a</span>
<span class="sd">    linear model to fit the response of multiple drugs using the expression of a gene, while the drugs are</span>
<span class="sd">    one-hot encoded. The p-value associated with the coefficient of gene expression is used as the prediction</span>
<span class="sd">    power measure, according to which num_predictive_gene genes will be selected. Then, among the predictive</span>
<span class="sd">    genes, num_generalizable_gene generalizable genes will be  selected.</span>
<span class="sd">    (3) If union_of_single_drug_selection is False and prediction_power_measure is &#39;pearson&#39; or &#39;mutual_info&#39;,</span>
<span class="sd">    for each drug this functions ranks the genes according to their power of predicting the</span>
<span class="sd">    response of the drug. The union of an equal number of predictive genes for every drug will be generated,</span>
<span class="sd">    and its size must be at least num_predictive_gene. Then, num_generalizable_gene generalizable genes</span>
<span class="sd">    will be selected.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    source_data: pandas data frame of gene expressions of tumors, for which drug response is known. Its size is</span>
<span class="sd">        [n_source_samples, n_features].</span>
<span class="sd">    target_data: pandas data frame of gene expressions of tumors, for which drug response needs to be predicted.</span>
<span class="sd">        Its size is [n_target_samples, n_features]. source_data and target_data have the same set</span>
<span class="sd">        of features and the orders of features must match.</span>
<span class="sd">    drug_response_data: pandas data frame of drug response that must include a column of drug response values,</span>
<span class="sd">        a column of tumor IDs, and a column of drug IDs.</span>
<span class="sd">    drug_response_col: non-negative integer or string. If integer, it is the column index of drug response in</span>
<span class="sd">        drug_response_data. If string, it is the column name of drug response.</span>
<span class="sd">    tumor_col: non-negative integer or string. If integer, it is the column index of tumor IDs in drug_response_data.</span>
<span class="sd">        If string, it is the column name of tumor IDs.</span>
<span class="sd">    drug_col: non-negative integer or string. If integer, it is the column index of drugs in drug_response_data.</span>
<span class="sd">        If string, it is the column name of drugs.</span>
<span class="sd">    prediction_power_measure: string. &#39;pearson&#39; uses the absolute value of Pearson correlation coefficient to</span>
<span class="sd">        measure prediction power of a gene; &#39;mutual_info&#39; uses the mutual information to measure prediction power</span>
<span class="sd">        of a gene; &#39;lm&#39; uses the linear regression model to select predictive genes for multiple drugs. Default is &#39;lm&#39;.</span>
<span class="sd">    num_predictive_gene: positive integer indicating the number of predictive genes to be selected.</span>
<span class="sd">    generalization_power_measure: string. &#39;pearson&#39; indicates the Pearson correlation coefficient;</span>
<span class="sd">        &#39;ccc&#39; indicates the concordance correlation coefficient. Default is &#39;ccc&#39;.</span>
<span class="sd">    num_generalizable_gene: positive integer indicating the number of generalizable genes to be selected.</span>
<span class="sd">    union_of_single_drug_selection: boolean, indicating whether the final gene set should be the union of genes</span>
<span class="sd">        selected for every drug.</span>

<span class="sd">    Returns:</span>
<span class="sd">    --------</span>
<span class="sd">    indices: 1-D numpy array containing the indices of selected genes.</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">drug_response_col</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">drug_response_col</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">drug_response_data</span><span class="o">.</span><span class="n">columns</span> <span class="o">==</span> <span class="n">drug_response_col</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">tumor_col</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">tumor_col</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">drug_response_data</span><span class="o">.</span><span class="n">columns</span> <span class="o">==</span> <span class="n">tumor_col</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">drug_col</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">drug_col</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">drug_response_data</span><span class="o">.</span><span class="n">columns</span> <span class="o">==</span> <span class="n">drug_col</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

    <span class="n">drug_response_data</span> <span class="o">=</span> <span class="n">drug_response_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">drug_response_data</span> <span class="o">=</span> <span class="n">drug_response_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">drug_response_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">tumor_col</span><span class="p">],</span>
                                                                  <span class="n">source_data</span><span class="o">.</span><span class="n">index</span><span class="p">))[</span><span class="mi">0</span><span class="p">],</span> <span class="p">:]</span>
    <span class="n">drugs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">drug_response_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">drug_col</span><span class="p">])</span>

    <span class="n">source_data</span> <span class="o">=</span> <span class="n">source_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">source_data</span> <span class="o">=</span> <span class="n">source_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">source_data</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">drug_response_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">tumor_col</span><span class="p">]))[</span><span class="mi">0</span><span class="p">],</span> <span class="p">:]</span>

    <span class="n">source_std_id</span> <span class="o">=</span> <span class="n">select_features_by_variation</span><span class="p">(</span><span class="n">source_data</span><span class="p">,</span> <span class="n">variation_measure</span><span class="o">=</span><span class="s1">&#39;std&#39;</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">0.00000001</span><span class="p">)</span>
    <span class="n">target_std_id</span> <span class="o">=</span> <span class="n">select_features_by_variation</span><span class="p">(</span><span class="n">target_data</span><span class="p">,</span> <span class="n">variation_measure</span><span class="o">=</span><span class="s1">&#39;std&#39;</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">0.00000001</span><span class="p">)</span>
    <span class="n">std_id</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="n">source_std_id</span><span class="p">,</span> <span class="n">target_std_id</span><span class="p">))</span>
    <span class="n">source_data</span> <span class="o">=</span> <span class="n">source_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">std_id</span><span class="p">]</span>
    <span class="n">target_data</span> <span class="o">=</span> <span class="n">target_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">target_data</span> <span class="o">=</span> <span class="n">target_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">std_id</span><span class="p">]</span>

    <span class="n">num_predictive_gene</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">num_predictive_gene</span><span class="p">,</span> <span class="n">source_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>

    <span class="k">if</span> <span class="n">union_of_single_drug_selection</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">prediction_power_measure</span> <span class="o">!=</span> <span class="s1">&#39;pearson&#39;</span> <span class="ow">and</span> <span class="n">prediction_power_measure</span> <span class="o">!=</span> <span class="s1">&#39;mutual_info&#39;</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;pearson or mutual_info must be used as prediction_power_measure for taking the union of selected genes of every drugs&#39;</span><span class="p">)</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">gid1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([])</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">drugs</span><span class="p">:</span>
            <span class="n">idd</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">drug_response_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">drug_col</span><span class="p">]</span> <span class="o">==</span> <span class="n">d</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">response_d</span> <span class="o">=</span> <span class="n">drug_response_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">idd</span><span class="p">,</span> <span class="p">:]</span>
            <span class="n">gid2</span> <span class="o">=</span> <span class="n">coxen_single_drug_gene_selection</span><span class="p">(</span><span class="n">source_data</span><span class="p">,</span> <span class="n">target_data</span><span class="p">,</span> <span class="n">response_d</span><span class="p">,</span> <span class="n">drug_response_col</span><span class="p">,</span> <span class="n">tumor_col</span><span class="p">,</span>
                <span class="n">prediction_power_measure</span><span class="p">,</span> <span class="n">num_predictive_gene</span><span class="p">,</span> <span class="n">generalization_power_measure</span><span class="p">,</span> <span class="n">num_generalizable_gene</span><span class="p">)</span>
            <span class="n">gid1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">union1d</span><span class="p">(</span><span class="n">gid1</span><span class="p">,</span> <span class="n">gid2</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">std_id</span><span class="p">[</span><span class="n">gid1</span><span class="p">])</span>

    <span class="k">if</span> <span class="n">prediction_power_measure</span> <span class="o">==</span> <span class="s1">&#39;lm&#39;</span><span class="p">:</span>
        <span class="n">pvalue</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">((</span><span class="n">source_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="mi">1</span><span class="p">))</span>
        <span class="n">pvalue</span><span class="o">.</span><span class="n">fill</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
        <span class="n">drug_m</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">identity</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">drugs</span><span class="p">))</span>
        <span class="n">drug_m</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">drug_m</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">drugs</span><span class="p">)</span>
        <span class="n">drug_sample</span> <span class="o">=</span> <span class="n">drug_m</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">drug_response_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">drug_col</span><span class="p">],</span> <span class="p">:]</span><span class="o">.</span><span class="n">values</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">source_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
            <span class="n">ge_sample</span> <span class="o">=</span> <span class="n">source_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">drug_response_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">tumor_col</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span>
            <span class="n">sample</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">ge_sample</span><span class="p">,</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ge_sample</span><span class="p">),</span> <span class="mi">1</span><span class="p">)),</span> <span class="n">drug_sample</span><span class="p">))</span>
            <span class="n">sample</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">add_constant</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
            <span class="n">mod</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">OLS</span><span class="p">(</span><span class="n">drug_response_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">drug_response_col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">sample</span><span class="p">)</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">res</span> <span class="o">=</span> <span class="n">mod</span><span class="o">.</span><span class="n">fit</span><span class="p">()</span>
                <span class="n">pvalue</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">pvalues</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="n">pvalue</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>

        <span class="n">gid1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">pvalue</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">])[:</span><span class="n">num_predictive_gene</span><span class="p">]</span>

    <span class="k">elif</span> <span class="n">prediction_power_measure</span> <span class="o">==</span> <span class="s1">&#39;pearson&#39;</span> <span class="ow">or</span> <span class="n">prediction_power_measure</span> <span class="o">==</span> <span class="s1">&#39;mutual_info&#39;</span><span class="p">:</span>
        <span class="n">gene_rank</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">drugs</span><span class="p">),</span> <span class="n">source_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
        <span class="n">gene_rank</span><span class="o">.</span><span class="n">fill</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
        <span class="n">gene_rank</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">gene_rank</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">drugs</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">drugs</span><span class="p">)):</span>
            <span class="n">idd</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">drug_response_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">drug_col</span><span class="p">]</span> <span class="o">==</span> <span class="n">drugs</span><span class="p">[</span><span class="n">d</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">response_d</span> <span class="o">=</span> <span class="n">drug_response_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">idd</span><span class="p">,</span> <span class="p">:]</span>
            <span class="n">temp_rank</span> <span class="o">=</span> <span class="n">coxen_single_drug_gene_selection</span><span class="p">(</span><span class="n">source_data</span><span class="p">,</span> <span class="n">target_data</span><span class="p">,</span> <span class="n">response_d</span><span class="p">,</span>
                <span class="n">drug_response_col</span><span class="p">,</span> <span class="n">tumor_col</span><span class="p">,</span> <span class="n">prediction_power_measure</span><span class="p">,</span> <span class="n">num_predictive_gene</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                <span class="n">generalization_power_measure</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">num_generalizable_gene</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">multi_drug_mode</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">gene_rank</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">d</span><span class="p">,</span> <span class="p">:</span><span class="nb">len</span><span class="p">(</span><span class="n">temp_rank</span><span class="p">)]</span> <span class="o">=</span> <span class="n">temp_rank</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="n">num_predictive_gene</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">drugs</span><span class="p">))),</span> <span class="n">source_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
            <span class="n">gid1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">gene_rank</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="p">:</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">gene_rank</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">*</span><span class="n">i</span><span class="p">))[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:])</span>
            <span class="n">gid1</span> <span class="o">=</span> <span class="n">gid1</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">invert</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">gid1</span><span class="p">)))[</span><span class="mi">0</span><span class="p">]]</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gid1</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">num_predictive_gene</span><span class="p">:</span>
                <span class="k">break</span>
        <span class="n">gid1</span> <span class="o">=</span> <span class="n">gid1</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">)</span>

    <span class="c1"># keep only predictive genes for source and target data</span>
    <span class="n">source_data</span> <span class="o">=</span> <span class="n">source_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">gid1</span><span class="p">]</span>
    <span class="n">target_data</span> <span class="o">=</span> <span class="n">target_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">gid1</span><span class="p">]</span>
    <span class="n">num_generalizable_gene</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">num_generalizable_gene</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">gid1</span><span class="p">)))</span>

    <span class="c1"># perform the second step of COXEN approach to select generalizable genes among the predictive genes</span>
    <span class="n">gid2</span> <span class="o">=</span> <span class="n">generalization_feature_selection</span><span class="p">(</span><span class="n">source_data</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">target_data</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">generalization_power_measure</span><span class="p">,</span>
                                            <span class="n">num_generalizable_gene</span><span class="p">)</span>

    <span class="n">indices</span> <span class="o">=</span> <span class="n">std_id</span><span class="p">[</span><span class="n">gid1</span><span class="p">[</span><span class="n">gid2</span><span class="p">]]</span>

    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span></div>

<div class="viewcode-block" id="generate_gene_set_data"><a class="viewcode-back" href="../candle_lib/P1_utils.html#P1_utils.generate_gene_set_data">[docs]</a><span class="k">def</span> <span class="nf">generate_gene_set_data</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">gene_name_type</span><span class="o">=</span><span class="s1">&#39;entrez&#39;</span><span class="p">,</span> <span class="n">gene_set_category</span><span class="o">=</span><span class="s1">&#39;c6.all&#39;</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;mean&#39;</span><span class="p">,</span>
                           <span class="n">standardize</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">data_dir</span><span class="o">=</span><span class="s1">&#39;../../Data/examples/Gene_Sets/MSigDB.v7.0/&#39;</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    This function generates genomic data summarized at the gene set level.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    data: numpy array or pandas data frame of numeric values, with a shape of [n_samples, n_features].</span>
<span class="sd">    genes: 1-D array or list of gene names with a length of n_features. It indicates which gene a genomic</span>
<span class="sd">        feature belongs to.</span>
<span class="sd">    gene_name_type: string, indicating the type of gene name used in genes. &#39;entrez&#39; indicates Entrez gene ID and</span>
<span class="sd">        &#39;symbols&#39; indicates HGNC gene symbol. Default is &#39;symbols&#39;.</span>
<span class="sd">    gene_set_category: string, indicating the gene sets for which data will be calculated. &#39;c2.cgp&#39; indicates gene sets</span>
<span class="sd">        affected by chemical and genetic perturbations; &#39;c2.cp.biocarta&#39; indicates BioCarta gene sets; &#39;c2.cp.kegg&#39;</span>
<span class="sd">        indicates KEGG gene sets; &#39;c2.cp.pid&#39; indicates PID gene sets; &#39;c2.cp.reactome&#39; indicates Reactome gene sets;</span>
<span class="sd">        &#39;c5.bp&#39; indicates GO biological processes; &#39;c5.cc&#39; indicates GO cellular components; &#39;c5.mf&#39; indicates</span>
<span class="sd">        GO molecular functions; &#39;c6.all&#39; indicates oncogenic signatures. Default is &#39;c6.all&#39;.</span>
<span class="sd">    metric: string, indicating the way to calculate gene-set-level data. &#39;mean&#39; calculates the mean of gene</span>
<span class="sd">        features belonging to the same gene set. &#39;sum&#39; calculates the summation of gene features belonging</span>
<span class="sd">        to the same gene set. &#39;max&#39; calculates the maximum of gene features. &#39;min&#39; calculates the minimum</span>
<span class="sd">        of gene features. &#39;abs_mean&#39; calculates the mean of absolute values. &#39;abs_maximum&#39; calculates</span>
<span class="sd">        the maximum of absolute values. Default is &#39;mean&#39;.</span>
<span class="sd">    standardize: boolean, indicating whether to standardize features before calculation. Standardization transforms</span>
<span class="sd">        each feature to have a zero mean and a unit standard deviation.</span>

<span class="sd">    Returns:</span>
<span class="sd">    --------</span>
<span class="sd">    gene_set_data: a data frame of calculated gene-set-level data. Column names are the gene set names.</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="n">sample_name</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
        <span class="n">sample_name</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">index</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">values</span>
    <span class="k">elif</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Input data must be a numpy array or pandas data frame&#39;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">standardize</span><span class="p">:</span>
        <span class="n">scaler</span> <span class="o">=</span> <span class="n">StandardScaler</span><span class="p">()</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">scaler</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

    <span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">gene_name_type</span> <span class="o">==</span> <span class="s1">&#39;entrez&#39;</span><span class="p">:</span>
        <span class="n">gene_set_category</span> <span class="o">=</span> <span class="n">gene_set_category</span> <span class="o">+</span> <span class="s1">&#39;.v7.0.entrez.gmt&#39;</span>
    <span class="k">if</span> <span class="n">gene_name_type</span> <span class="o">==</span> <span class="s1">&#39;symbols&#39;</span><span class="p">:</span>
        <span class="n">gene_set_category</span> <span class="o">=</span> <span class="n">gene_set_category</span> <span class="o">+</span> <span class="s1">&#39;.v7.0.symbols.gmt&#39;</span>
    <span class="n">f</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">data_dir</span>  <span class="o">+</span> <span class="n">gene_set_category</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
    <span class="n">x</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
    <span class="n">gene_sets</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)):</span>
        <span class="n">temp</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">gene_sets</span><span class="p">[</span><span class="n">temp</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[</span><span class="mi">2</span><span class="p">:]</span>

    <span class="n">gene_set_data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">((</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_sets</span><span class="p">)))</span>
    <span class="n">gene_set_data</span><span class="o">.</span><span class="n">fill</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
    <span class="n">gene_set_names</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">gene_sets</span><span class="o">.</span><span class="n">keys</span><span class="p">()))</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_set_names</span><span class="p">)):</span>
        <span class="n">idi</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">gene_sets</span><span class="p">[</span><span class="n">gene_set_names</span><span class="p">[</span><span class="n">i</span><span class="p">]]))[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">idi</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">metric</span> <span class="o">==</span> <span class="s1">&#39;sum&#39;</span><span class="p">:</span>
                <span class="n">gene_set_data</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nansum</span><span class="p">(</span><span class="n">data</span><span class="p">[:,</span> <span class="n">idi</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">metric</span> <span class="o">==</span> <span class="s1">&#39;max&#39;</span><span class="p">:</span>
                <span class="n">gene_set_data</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmax</span><span class="p">(</span><span class="n">data</span><span class="p">[:,</span> <span class="n">idi</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">metric</span> <span class="o">==</span> <span class="s1">&#39;min&#39;</span><span class="p">:</span>
                <span class="n">gene_set_data</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmin</span><span class="p">(</span><span class="n">data</span><span class="p">[:,</span> <span class="n">idi</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">metric</span> <span class="o">==</span> <span class="s1">&#39;abs_mean&#39;</span><span class="p">:</span>
                <span class="n">gene_set_data</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">absolute</span><span class="p">(</span><span class="n">data</span><span class="p">[:,</span> <span class="n">idi</span><span class="p">]),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">metric</span> <span class="o">==</span> <span class="s1">&#39;abs_maximum&#39;</span><span class="p">:</span>
                <span class="n">gene_set_data</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmax</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">absolute</span><span class="p">(</span><span class="n">data</span><span class="p">[:,</span> <span class="n">idi</span><span class="p">]),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>  <span class="c1">#&#39;mean&#39;</span>
                <span class="n">gene_set_data</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">data</span><span class="p">[:,</span> <span class="n">idi</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">sample_name</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">gene_set_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">gene_set_data</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">gene_set_names</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">gene_set_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">gene_set_data</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">gene_set_names</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">sample_name</span><span class="p">)</span>
    <span class="n">keep_id</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">invert</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">isna</span><span class="p">(</span><span class="n">gene_set_data</span><span class="p">)),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">gene_set_data</span> <span class="o">=</span> <span class="n">gene_set_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">keep_id</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">gene_set_data</span></div>

<span class="c1">################### Auxiliary functions of ComBat start here ####################</span>



<div class="viewcode-block" id="design_mat"><a class="viewcode-back" href="../candle_lib/P1_utils.html#P1_utils.design_mat">[docs]</a><span class="k">def</span> <span class="nf">design_mat</span><span class="p">(</span><span class="n">mod</span><span class="p">,</span> <span class="n">numerical_covariates</span><span class="p">,</span> <span class="n">batch_levels</span><span class="p">):</span>
    <span class="c1"># require levels to make sure they are in the same order as we use in the</span>
    <span class="c1"># rest of the script.</span>
    <span class="n">design</span> <span class="o">=</span> <span class="n">patsy</span><span class="o">.</span><span class="n">dmatrix</span><span class="p">(</span><span class="s2">&quot;~ 0 + C(batch, levels=</span><span class="si">%s</span><span class="s2">)&quot;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">batch_levels</span><span class="p">),</span>
                                                  <span class="n">mod</span><span class="p">,</span> <span class="n">return_type</span><span class="o">=</span><span class="s2">&quot;dataframe&quot;</span><span class="p">)</span>

    <span class="n">mod</span> <span class="o">=</span> <span class="n">mod</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s2">&quot;batch&quot;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">numerical_covariates</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">numerical_covariates</span><span class="p">)</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;found </span><span class="si">%i</span><span class="s2"> batches</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">design</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="n">other_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">mod</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
                  <span class="k">if</span> <span class="ow">not</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">numerical_covariates</span><span class="p">]</span>
    <span class="n">factor_matrix</span> <span class="o">=</span> <span class="n">mod</span><span class="p">[</span><span class="n">other_cols</span><span class="p">]</span>
    <span class="n">design</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">((</span><span class="n">design</span><span class="p">,</span> <span class="n">factor_matrix</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">numerical_covariates</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;found </span><span class="si">%i</span><span class="s2"> numerical covariates...</span><span class="se">\n</span><span class="s2">&quot;</span>
                            <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">numerical_covariates</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">nC</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">numerical_covariates</span><span class="p">):</span>
            <span class="n">cname</span> <span class="o">=</span> <span class="n">mod</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="n">nC</span><span class="p">]</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="si">{0}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cname</span><span class="p">))</span>
            <span class="n">design</span><span class="p">[</span><span class="n">cname</span><span class="p">]</span> <span class="o">=</span> <span class="n">mod</span><span class="p">[</span><span class="n">mod</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="n">nC</span><span class="p">]]</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;found </span><span class="si">%i</span><span class="s2"> categorical variables:&quot;</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">other_cols</span><span class="p">))</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">other_cols</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">design</span></div>


<div class="viewcode-block" id="it_sol"><a class="viewcode-back" href="../candle_lib/P1_utils.html#P1_utils.it_sol">[docs]</a><span class="k">def</span> <span class="nf">it_sol</span><span class="p">(</span><span class="n">sdat</span><span class="p">,</span> <span class="n">g_hat</span><span class="p">,</span> <span class="n">d_hat</span><span class="p">,</span> <span class="n">g_bar</span><span class="p">,</span> <span class="n">t2</span><span class="p">,</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">conv</span><span class="o">=</span><span class="mf">0.0001</span><span class="p">):</span>
    <span class="n">n</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">sdat</span><span class="p">))</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">g_old</span> <span class="o">=</span> <span class="n">g_hat</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">d_old</span> <span class="o">=</span> <span class="n">d_hat</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="n">change</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">while</span> <span class="n">change</span> <span class="o">&gt;</span> <span class="n">conv</span><span class="p">:</span>
        <span class="c1"># print g_hat.shape, g_bar.shape, t2.shape</span>
        <span class="n">g_new</span> <span class="o">=</span> <span class="n">postmean</span><span class="p">(</span><span class="n">g_hat</span><span class="p">,</span> <span class="n">g_bar</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">d_old</span><span class="p">,</span> <span class="n">t2</span><span class="p">)</span>
        <span class="n">sum2</span> <span class="o">=</span> <span class="p">((</span><span class="n">sdat</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">g_new</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="n">g_new</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">1</span><span class="p">)),</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span> <span class="n">sdat</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))))</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
            <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">d_new</span> <span class="o">=</span> <span class="n">postvar</span><span class="p">(</span><span class="n">sum2</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span>

        <span class="n">change</span> <span class="o">=</span> <span class="nb">max</span><span class="p">((</span><span class="nb">abs</span><span class="p">(</span><span class="n">g_new</span> <span class="o">-</span> <span class="n">g_old</span><span class="p">)</span> <span class="o">/</span> <span class="n">g_old</span><span class="p">)</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span> <span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">d_new</span> <span class="o">-</span> <span class="n">d_old</span><span class="p">)</span> <span class="o">/</span> <span class="n">d_old</span><span class="p">)</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>
        <span class="n">g_old</span> <span class="o">=</span> <span class="n">g_new</span>  <span class="c1"># .copy()</span>
        <span class="n">d_old</span> <span class="o">=</span> <span class="n">d_new</span>  <span class="c1"># .copy()</span>
        <span class="n">count</span> <span class="o">=</span> <span class="n">count</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="n">adjust</span> <span class="o">=</span> <span class="p">(</span><span class="n">g_new</span><span class="p">,</span> <span class="n">d_new</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">adjust</span></div>


<div class="viewcode-block" id="aprior"><a class="viewcode-back" href="../candle_lib/P1_utils.html#P1_utils.aprior">[docs]</a><span class="k">def</span> <span class="nf">aprior</span><span class="p">(</span><span class="n">gamma_hat</span><span class="p">):</span>
    <span class="n">m</span> <span class="o">=</span> <span class="n">gamma_hat</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
    <span class="n">s2</span> <span class="o">=</span> <span class="n">gamma_hat</span><span class="o">.</span><span class="n">var</span><span class="p">()</span>
    <span class="k">return</span> <span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">s2</span> <span class="o">+</span> <span class="n">m</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="n">s2</span></div>


<div class="viewcode-block" id="bprior"><a class="viewcode-back" href="../candle_lib/P1_utils.html#P1_utils.bprior">[docs]</a><span class="k">def</span> <span class="nf">bprior</span><span class="p">(</span><span class="n">gamma_hat</span><span class="p">):</span>
    <span class="n">m</span> <span class="o">=</span> <span class="n">gamma_hat</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
    <span class="n">s2</span> <span class="o">=</span> <span class="n">gamma_hat</span><span class="o">.</span><span class="n">var</span><span class="p">()</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">m</span> <span class="o">*</span> <span class="n">s2</span> <span class="o">+</span> <span class="n">m</span> <span class="o">**</span> <span class="mi">3</span><span class="p">)</span> <span class="o">/</span> <span class="n">s2</span></div>


<div class="viewcode-block" id="postmean"><a class="viewcode-back" href="../candle_lib/P1_utils.html#P1_utils.postmean">[docs]</a><span class="k">def</span> <span class="nf">postmean</span><span class="p">(</span><span class="n">g_hat</span><span class="p">,</span> <span class="n">g_bar</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">d_star</span><span class="p">,</span> <span class="n">t2</span><span class="p">):</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">t2</span> <span class="o">*</span> <span class="n">n</span> <span class="o">*</span> <span class="n">g_hat</span> <span class="o">+</span> <span class="n">d_star</span> <span class="o">*</span> <span class="n">g_bar</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">t2</span> <span class="o">*</span> <span class="n">n</span> <span class="o">+</span> <span class="n">d_star</span><span class="p">)</span></div>


<div class="viewcode-block" id="postvar"><a class="viewcode-back" href="../candle_lib/P1_utils.html#P1_utils.postvar">[docs]</a><span class="k">def</span> <span class="nf">postvar</span><span class="p">(</span><span class="n">sum2</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">):</span>
    <span class="k">return</span> <span class="p">(</span><span class="mf">0.5</span> <span class="o">*</span> <span class="n">sum2</span> <span class="o">+</span> <span class="n">b</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">n</span> <span class="o">/</span> <span class="mf">2.0</span> <span class="o">+</span> <span class="n">a</span> <span class="o">-</span> <span class="mf">1.0</span><span class="p">)</span></div>



<span class="c1">################### Auxiliary functions of ComBat end here ####################</span>

<div class="viewcode-block" id="combat_batch_effect_removal"><a class="viewcode-back" href="../candle_lib/P1_utils.html#P1_utils.combat_batch_effect_removal">[docs]</a><span class="k">def</span> <span class="nf">combat_batch_effect_removal</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">batch_labels</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">numerical_covariates</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    This function corrects for batch effect in data.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    data: pandas data frame of numeric values, with a size of (n_features, n_samples)</span>
<span class="sd">    batch_labels: pandas series, with a length of n_samples. It should provide the batch labels of samples.</span>
<span class="sd">        Its indices are the same as the column names (sample names) in &quot;data&quot;.</span>
<span class="sd">    model: an object of patsy.design_info.DesignMatrix. It is a design matrix describing the covariate</span>
<span class="sd">        information on the samples that could cause batch effects. If not provided, this function</span>
<span class="sd">        will attempt to coarsely correct just based on the information provided in &quot;batch&quot;.</span>
<span class="sd">    numerical_covariates: a list of the names of covariates in &quot;model&quot; that are numerical rather than</span>
<span class="sd">        categorical.</span>

<span class="sd">    Returns:</span>
<span class="sd">    --------</span>
<span class="sd">    corrected : pandas data frame of numeric values, with a size of (n_features, n_samples). It is</span>
<span class="sd">        the data with batch effects corrected.</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">numerical_covariates</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">numerical_covariates</span> <span class="o">=</span> <span class="p">[</span><span class="n">numerical_covariates</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">numerical_covariates</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">numerical_covariates</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">if</span> <span class="n">model</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
        <span class="n">model</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">batch_labels</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;batch&#39;</span><span class="p">:</span> <span class="n">batch_labels</span><span class="p">})</span>

    <span class="n">batch_items</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;batch&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">groups</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
    <span class="n">batch_levels</span> <span class="o">=</span> <span class="p">[</span><span class="n">k</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">batch_items</span><span class="p">]</span>
    <span class="n">batch_info</span> <span class="o">=</span> <span class="p">[</span><span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">batch_items</span><span class="p">]</span>
    <span class="n">n_batch</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">batch_info</span><span class="p">)</span>
    <span class="n">n_batches</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">len</span><span class="p">(</span><span class="n">v</span><span class="p">)</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">batch_info</span><span class="p">])</span>
    <span class="n">n_array</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">n_batches</span><span class="p">))</span>

    <span class="c1"># drop intercept</span>
    <span class="n">drop_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">cname</span> <span class="k">for</span> <span class="n">cname</span><span class="p">,</span> <span class="n">inter</span> <span class="ow">in</span> <span class="p">((</span><span class="n">model</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">())</span><span class="o">.</span><span class="n">iteritems</span><span class="p">()</span> <span class="k">if</span> <span class="n">inter</span> <span class="o">==</span> <span class="kc">True</span><span class="p">]</span>
    <span class="n">drop_idxs</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">cdrop</span><span class="p">)</span> <span class="k">for</span> <span class="n">cdrop</span> <span class="ow">in</span> <span class="n">drop_cols</span><span class="p">]</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">model</span><span class="p">[[</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">model</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">drop_cols</span><span class="p">]]</span>
    <span class="n">numerical_covariates</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">c</span>
                            <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">numerical_covariates</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">drop_cols</span><span class="p">]</span>

    <span class="n">design</span> <span class="o">=</span> <span class="n">design_mat</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">numerical_covariates</span><span class="p">,</span> <span class="n">batch_levels</span><span class="p">)</span>

    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;Standardizing Data across genes.</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">B_hat</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">la</span><span class="o">.</span><span class="n">inv</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">design</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">design</span><span class="p">)),</span> <span class="n">design</span><span class="o">.</span><span class="n">T</span><span class="p">),</span> <span class="n">data</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
    <span class="n">grand_mean</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">((</span><span class="n">n_batches</span> <span class="o">/</span> <span class="n">n_array</span><span class="p">)</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">B_hat</span><span class="p">[:</span><span class="n">n_batch</span><span class="p">,</span> <span class="p">:])</span>
    <span class="n">var_pooled</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(((</span><span class="n">data</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">design</span><span class="p">,</span> <span class="n">B_hat</span><span class="p">)</span><span class="o">.</span><span class="n">T</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="nb">int</span><span class="p">(</span><span class="n">n_array</span><span class="p">),</span> <span class="mi">1</span><span class="p">))</span> <span class="o">/</span> <span class="nb">int</span><span class="p">(</span><span class="n">n_array</span><span class="p">))</span>

    <span class="n">stand_mean</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">grand_mean</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">grand_mean</span><span class="p">),</span> <span class="mi">1</span><span class="p">)),</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">n_array</span><span class="p">))))</span>
    <span class="n">tmp</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">design</span><span class="o">.</span><span class="n">copy</span><span class="p">())</span>
    <span class="n">tmp</span><span class="p">[:,</span> <span class="p">:</span><span class="n">n_batch</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">stand_mean</span> <span class="o">+=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">tmp</span><span class="p">,</span> <span class="n">B_hat</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>

    <span class="n">s_data</span> <span class="o">=</span> <span class="p">((</span><span class="n">data</span> <span class="o">-</span> <span class="n">stand_mean</span><span class="p">)</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">var_pooled</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">n_array</span><span class="p">)))))</span>

    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;Fitting L/S model and finding priors</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">batch_design</span> <span class="o">=</span> <span class="n">design</span><span class="p">[</span><span class="n">design</span><span class="o">.</span><span class="n">columns</span><span class="p">[:</span><span class="n">n_batch</span><span class="p">]]</span>
    <span class="n">gamma_hat</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">la</span><span class="o">.</span><span class="n">inv</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">batch_design</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">batch_design</span><span class="p">)),</span> <span class="n">batch_design</span><span class="o">.</span><span class="n">T</span><span class="p">),</span> <span class="n">s_data</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>

    <span class="n">delta_hat</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">batch_idxs</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">batch_info</span><span class="p">):</span>
        <span class="n">delta_hat</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">s_data</span><span class="p">[</span><span class="n">batch_idxs</span><span class="p">]</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>

    <span class="n">gamma_bar</span> <span class="o">=</span> <span class="n">gamma_hat</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">t2</span> <span class="o">=</span> <span class="n">gamma_hat</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">a_prior</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="n">aprior</span><span class="p">,</span> <span class="n">delta_hat</span><span class="p">))</span>
    <span class="n">b_prior</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="n">bprior</span><span class="p">,</span> <span class="n">delta_hat</span><span class="p">))</span>

    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;Finding parametric adjustments</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">gamma_star</span><span class="p">,</span> <span class="n">delta_star</span> <span class="o">=</span> <span class="p">[],</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">batch_idxs</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">batch_info</span><span class="p">):</span>
        <span class="n">temp</span> <span class="o">=</span> <span class="n">it_sol</span><span class="p">(</span><span class="n">s_data</span><span class="p">[</span><span class="n">batch_idxs</span><span class="p">],</span> <span class="n">gamma_hat</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                      <span class="n">delta_hat</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">gamma_bar</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">t2</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">a_prior</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">b_prior</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>

        <span class="n">gamma_star</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">temp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">delta_star</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">temp</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

    <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;Adjusting data</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">bayesdata</span> <span class="o">=</span> <span class="n">s_data</span>
    <span class="n">gamma_star</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">gamma_star</span><span class="p">)</span>
    <span class="n">delta_star</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">delta_star</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">batch_idxs</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">batch_info</span><span class="p">):</span>
        <span class="n">dsq</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">delta_star</span><span class="p">[</span><span class="n">j</span><span class="p">,</span> <span class="p">:])</span>
        <span class="n">dsq</span> <span class="o">=</span> <span class="n">dsq</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">dsq</span><span class="p">),</span> <span class="mi">1</span><span class="p">))</span>
        <span class="n">denom</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">dsq</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span> <span class="n">n_batches</span><span class="p">[</span><span class="n">j</span><span class="p">])))</span>
        <span class="n">numer</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">bayesdata</span><span class="p">[</span><span class="n">batch_idxs</span><span class="p">]</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">batch_design</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">batch_idxs</span><span class="p">],</span> <span class="n">gamma_star</span><span class="p">)</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>

        <span class="n">bayesdata</span><span class="p">[</span><span class="n">batch_idxs</span><span class="p">]</span> <span class="o">=</span> <span class="n">numer</span> <span class="o">/</span> <span class="n">denom</span>

    <span class="n">vpsq</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">var_pooled</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">var_pooled</span><span class="p">),</span> <span class="mi">1</span><span class="p">))</span>
    <span class="n">bayesdata</span> <span class="o">=</span> <span class="n">bayesdata</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">vpsq</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">n_array</span><span class="p">))))</span> <span class="o">+</span> <span class="n">stand_mean</span>

    <span class="k">return</span> <span class="n">bayesdata</span></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, ANL.gov

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>