

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>P1_utils module &mdash; CANDLE  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="profiling_utils module" href="profiling_utils.html" />
    <link rel="prev" title="keras_utils module" href="keras_utils.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> CANDLE
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../readme.html">What is CANDLE?</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/index.html">Tutorials</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../candle_lib.html">CANDLE Library API</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="candle_keras.html">candle_keras package</a></li>
<li class="toctree-l2"><a class="reference internal" href="data_preprocessing_utils.html">data_preprocessing_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="data_utils.html">data_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="default_utils.html">default_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="exp_logger.html">exp_logger module</a></li>
<li class="toctree-l2"><a class="reference internal" href="feature_selection_utils.html">feature_selection_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="file_utils.html">file_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="generic_utils.html">generic_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="keras_utils.html">keras_utils module</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">P1_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="profiling_utils.html">profiling_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pytorch_utils.html">pytorch_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="solr_keras.html">solr_keras module</a></li>
<li class="toctree-l2"><a class="reference internal" href="uq_utils.html">uq_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="uq_keras_utils.html">uq_keras_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="viz_utils.html">viz_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="clr_keras_utils.html">clr_keras_utils module</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../contribute.html">How to Contribute</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">CANDLE</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../candle_lib.html">CANDLE Library API</a> &raquo;</li>
        
      <li>P1_utils module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/candle_lib/P1_utils.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-P1_utils">
<span id="p1-utils-module"></span><h1>P1_utils module<a class="headerlink" href="#module-P1_utils" title="Permalink to this headline">¶</a></h1>
<dl class="py function">
<dt id="P1_utils.aprior">
<code class="sig-prename descclassname"><span class="pre">P1_utils.</span></code><code class="sig-name descname"><span class="pre">aprior</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gamma_hat</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/P1_utils.html#aprior"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#P1_utils.aprior" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="P1_utils.bprior">
<code class="sig-prename descclassname"><span class="pre">P1_utils.</span></code><code class="sig-name descname"><span class="pre">bprior</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gamma_hat</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/P1_utils.html#bprior"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#P1_utils.bprior" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="P1_utils.calculate_concordance_correlation_coefficient">
<code class="sig-prename descclassname"><span class="pre">P1_utils.</span></code><code class="sig-name descname"><span class="pre">calculate_concordance_correlation_coefficient</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/P1_utils.html#calculate_concordance_correlation_coefficient"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#P1_utils.calculate_concordance_correlation_coefficient" title="Permalink to this definition">¶</a></dt>
<dd><p>This function calculates the concordance correlation coefficient between two input 1-D numpy arrays.</p>
<p>u: 1-D numpy array of a variable
v: 1-D numpy array of a variable</p>
<p>ccc: a numeric value of concordance correlation coefficient between the two input variables.</p>
</dd></dl>

<dl class="py function">
<dt id="P1_utils.combat_batch_effect_removal">
<code class="sig-prename descclassname"><span class="pre">P1_utils.</span></code><code class="sig-name descname"><span class="pre">combat_batch_effect_removal</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">numerical_covariates</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/P1_utils.html#combat_batch_effect_removal"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#P1_utils.combat_batch_effect_removal" title="Permalink to this definition">¶</a></dt>
<dd><p>This function corrects for batch effect in data.</p>
<p>data: pandas data frame of numeric values, with a size of (n_features, n_samples)
batch_labels: pandas series, with a length of n_samples. It should provide the batch labels of samples.</p>
<blockquote>
<div><p>Its indices are the same as the column names (sample names) in “data”.</p>
</div></blockquote>
<dl class="simple">
<dt>model: an object of patsy.design_info.DesignMatrix. It is a design matrix describing the covariate</dt><dd><p>information on the samples that could cause batch effects. If not provided, this function
will attempt to coarsely correct just based on the information provided in “batch”.</p>
</dd>
<dt>numerical_covariates: a list of the names of covariates in “model” that are numerical rather than</dt><dd><p>categorical.</p>
</dd>
</dl>
<dl class="simple">
<dt>corrected<span class="classifier">pandas data frame of numeric values, with a size of (n_features, n_samples). It is</span></dt><dd><p>the data with batch effects corrected.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="P1_utils.coxen_multi_drug_gene_selection">
<code class="sig-prename descclassname"><span class="pre">P1_utils.</span></code><code class="sig-name descname"><span class="pre">coxen_multi_drug_gene_selection</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drug_response_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drug_response_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tumor_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drug_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prediction_power_measure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'lm'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_predictive_gene</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">generalization_power_measure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ccc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_generalizable_gene</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">union_of_single_drug_selection</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/P1_utils.html#coxen_multi_drug_gene_selection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#P1_utils.coxen_multi_drug_gene_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>This function uses the COXEN approach to select genes for predicting the response of multiple drugs.
It assumes no missing data exist. It works in three modes.
(1) If union_of_single_drug_selection is True, prediction_power_measure must be either ‘pearson’ or ‘mutual_info’.
This functions runs coxen_single_drug_gene_selection for every drug with the parameter setting and takes the
union of the selected genes of every drug as the output. The size of the selected gene set may be larger than
num_generalizable_gene.
(2) If union_of_single_drug_selection is False and prediction_power_measure is ‘lm’, this function uses a
linear model to fit the response of multiple drugs using the expression of a gene, while the drugs are
one-hot encoded. The p-value associated with the coefficient of gene expression is used as the prediction
power measure, according to which num_predictive_gene genes will be selected. Then, among the predictive
genes, num_generalizable_gene generalizable genes will be  selected.
(3) If union_of_single_drug_selection is False and prediction_power_measure is ‘pearson’ or ‘mutual_info’,
for each drug this functions ranks the genes according to their power of predicting the
response of the drug. The union of an equal number of predictive genes for every drug will be generated,
and its size must be at least num_predictive_gene. Then, num_generalizable_gene generalizable genes
will be selected.</p>
<dl class="simple">
<dt>source_data: pandas data frame of gene expressions of tumors, for which drug response is known. Its size is</dt><dd><p>[n_source_samples, n_features].</p>
</dd>
<dt>target_data: pandas data frame of gene expressions of tumors, for which drug response needs to be predicted.</dt><dd><p>Its size is [n_target_samples, n_features]. source_data and target_data have the same set
of features and the orders of features must match.</p>
</dd>
<dt>drug_response_data: pandas data frame of drug response that must include a column of drug response values,</dt><dd><p>a column of tumor IDs, and a column of drug IDs.</p>
</dd>
<dt>drug_response_col: non-negative integer or string. If integer, it is the column index of drug response in</dt><dd><p>drug_response_data. If string, it is the column name of drug response.</p>
</dd>
<dt>tumor_col: non-negative integer or string. If integer, it is the column index of tumor IDs in drug_response_data.</dt><dd><p>If string, it is the column name of tumor IDs.</p>
</dd>
<dt>drug_col: non-negative integer or string. If integer, it is the column index of drugs in drug_response_data.</dt><dd><p>If string, it is the column name of drugs.</p>
</dd>
<dt>prediction_power_measure: string. ‘pearson’ uses the absolute value of Pearson correlation coefficient to</dt><dd><p>measure prediction power of a gene; ‘mutual_info’ uses the mutual information to measure prediction power
of a gene; ‘lm’ uses the linear regression model to select predictive genes for multiple drugs. Default is ‘lm’.</p>
</dd>
</dl>
<p>num_predictive_gene: positive integer indicating the number of predictive genes to be selected.
generalization_power_measure: string. ‘pearson’ indicates the Pearson correlation coefficient;</p>
<blockquote>
<div><p>‘ccc’ indicates the concordance correlation coefficient. Default is ‘ccc’.</p>
</div></blockquote>
<p>num_generalizable_gene: positive integer indicating the number of generalizable genes to be selected.
union_of_single_drug_selection: boolean, indicating whether the final gene set should be the union of genes</p>
<blockquote>
<div><p>selected for every drug.</p>
</div></blockquote>
<p>indices: 1-D numpy array containing the indices of selected genes.</p>
</dd></dl>

<dl class="py function">
<dt id="P1_utils.coxen_single_drug_gene_selection">
<code class="sig-prename descclassname"><span class="pre">P1_utils.</span></code><code class="sig-name descname"><span class="pre">coxen_single_drug_gene_selection</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drug_response_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drug_response_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tumor_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prediction_power_measure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pearson'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_predictive_gene</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">generalization_power_measure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ccc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_generalizable_gene</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multi_drug_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/P1_utils.html#coxen_single_drug_gene_selection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#P1_utils.coxen_single_drug_gene_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>This function selects genes for drug response prediction using the COXEN approach. The COXEN approach is
designed for selecting genes to predict the response of tumor cells to a specific drug. This function
assumes no missing data exist.</p>
<dl class="simple">
<dt>source_data: pandas data frame of gene expressions of tumors, for which drug response is known. Its size is</dt><dd><p>[n_source_samples, n_features].</p>
</dd>
<dt>target_data: pandas data frame of gene expressions of tumors, for which drug response needs to be predicted.</dt><dd><p>Its size is [n_target_samples, n_features]. source_data and target_data have the same set
of features and the orders of features must match.</p>
</dd>
<dt>drug_response_data: pandas data frame of drug response values for a drug. It must include a column of drug</dt><dd><p>response values and a column of tumor IDs.</p>
</dd>
<dt>drug_response_col: non-negative integer or string. If integer, it is the column index of drug response in</dt><dd><p>drug_response_data. If string, it is the column name of drug response.</p>
</dd>
<dt>tumor_col: non-negative integer or string. If integer, it is the column index of tumor IDs in drug_response_data.</dt><dd><p>If string, it is the column name of tumor IDs.</p>
</dd>
<dt>prediction_power_measure: string. ‘pearson’ uses the absolute value of Pearson correlation coefficient to</dt><dd><p>measure prediction power of gene; ‘mutual_info’ uses the mutual information to measure prediction power
of gene. Default is ‘pearson’.</p>
</dd>
</dl>
<p>num_predictive_gene: positive integer indicating the number of predictive genes to be selected.
generalization_power_measure: string. ‘pearson’ indicates the Pearson correlation coefficient;</p>
<blockquote>
<div><p>‘ccc’ indicates the concordance correlation coefficient. Default is ‘ccc’.</p>
</div></blockquote>
<p>num_generalizable_gene: positive integer indicating the number of generalizable genes to be selected.
multi_drug_mode: boolean, indicating whether the function runs as an auxiliary function of COXEN</p>
<blockquote>
<div><p>gene selection for multiple drugs. Default is False.</p>
</div></blockquote>
<p>indices: 1-D numpy array containing the indices of selected genes, if multi_drug_mode is False;
1-D numpy array of indices of sorting all genes according to their prediction power, if multi_drug_mode is True.</p>
</dd></dl>

<dl class="py function">
<dt id="P1_utils.design_mat">
<code class="sig-prename descclassname"><span class="pre">P1_utils.</span></code><code class="sig-name descname"><span class="pre">design_mat</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mod</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">numerical_covariates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_levels</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/P1_utils.html#design_mat"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#P1_utils.design_mat" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="P1_utils.generalization_feature_selection">
<code class="sig-prename descclassname"><span class="pre">P1_utils.</span></code><code class="sig-name descname"><span class="pre">generalization_feature_selection</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">measure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/P1_utils.html#generalization_feature_selection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#P1_utils.generalization_feature_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>This function uses the Pearson correlation coefficient to select the features that are generalizable
between data1 and data2.</p>
<p>data1: 2D numpy array of the first dataset with a size of (n_samples_1, n_features)
data2: 2D numpy array of the second dataset with a size of (n_samples_2, n_features)
measure: string. ‘pearson’ indicates the Pearson correlation coefficient;</p>
<blockquote>
<div><p>‘ccc’ indicates the concordance correlation coefficient. Default is ‘pearson’.</p>
</div></blockquote>
<dl class="simple">
<dt>cutoff: a positive number for selecting generalizable features. If cutoff &lt; 1, this function selects</dt><dd><p>the features with a correlation coefficient &gt;= cutoff. If cutoff &gt;= 1, it must be an
integer indicating the number of features to be selected based on correlation coefficient.</p>
</dd>
</dl>
<p>fid: 1-D numpy array containing the indices of selected features.</p>
</dd></dl>

<dl class="py function">
<dt id="P1_utils.generate_gene_set_data">
<code class="sig-prename descclassname"><span class="pre">P1_utils.</span></code><code class="sig-name descname"><span class="pre">generate_gene_set_data</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_name_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'entrez'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_set_category</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'c6.all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metric</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">standardize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'../../Data/examples/Gene_Sets/MSigDB.v7.0/'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/P1_utils.html#generate_gene_set_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#P1_utils.generate_gene_set_data" title="Permalink to this definition">¶</a></dt>
<dd><p>This function generates genomic data summarized at the gene set level.</p>
<p>data: numpy array or pandas data frame of numeric values, with a shape of [n_samples, n_features].
genes: 1-D array or list of gene names with a length of n_features. It indicates which gene a genomic</p>
<blockquote>
<div><p>feature belongs to.</p>
</div></blockquote>
<dl class="simple">
<dt>gene_name_type: string, indicating the type of gene name used in genes. ‘entrez’ indicates Entrez gene ID and</dt><dd><p>‘symbols’ indicates HGNC gene symbol. Default is ‘symbols’.</p>
</dd>
<dt>gene_set_category: string, indicating the gene sets for which data will be calculated. ‘c2.cgp’ indicates gene sets</dt><dd><p>affected by chemical and genetic perturbations; ‘c2.cp.biocarta’ indicates BioCarta gene sets; ‘c2.cp.kegg’
indicates KEGG gene sets; ‘c2.cp.pid’ indicates PID gene sets; ‘c2.cp.reactome’ indicates Reactome gene sets;
‘c5.bp’ indicates GO biological processes; ‘c5.cc’ indicates GO cellular components; ‘c5.mf’ indicates
GO molecular functions; ‘c6.all’ indicates oncogenic signatures. Default is ‘c6.all’.</p>
</dd>
<dt>metric: string, indicating the way to calculate gene-set-level data. ‘mean’ calculates the mean of gene</dt><dd><p>features belonging to the same gene set. ‘sum’ calculates the summation of gene features belonging
to the same gene set. ‘max’ calculates the maximum of gene features. ‘min’ calculates the minimum
of gene features. ‘abs_mean’ calculates the mean of absolute values. ‘abs_maximum’ calculates
the maximum of absolute values. Default is ‘mean’.</p>
</dd>
<dt>standardize: boolean, indicating whether to standardize features before calculation. Standardization transforms</dt><dd><p>each feature to have a zero mean and a unit standard deviation.</p>
</dd>
</dl>
<p>gene_set_data: a data frame of calculated gene-set-level data. Column names are the gene set names.</p>
</dd></dl>

<dl class="py function">
<dt id="P1_utils.it_sol">
<code class="sig-prename descclassname"><span class="pre">P1_utils.</span></code><code class="sig-name descname"><span class="pre">it_sol</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sdat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">g_hat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d_hat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">g_bar</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">t2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">a</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conv</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0001</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/P1_utils.html#it_sol"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#P1_utils.it_sol" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="P1_utils.postmean">
<code class="sig-prename descclassname"><span class="pre">P1_utils.</span></code><code class="sig-name descname"><span class="pre">postmean</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">g_hat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">g_bar</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d_star</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">t2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/P1_utils.html#postmean"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#P1_utils.postmean" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="P1_utils.postvar">
<code class="sig-prename descclassname"><span class="pre">P1_utils.</span></code><code class="sig-name descname"><span class="pre">postvar</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sum2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">a</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/P1_utils.html#postvar"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#P1_utils.postvar" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="profiling_utils.html" class="btn btn-neutral float-right" title="profiling_utils module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="keras_utils.html" class="btn btn-neutral float-left" title="keras_utils module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, ANL.gov

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>